include/mstream.h

```diff
@@ -0,0 +1,111 @@
+#pragma once
+
+#include <streambuf>
+#include <istream>
+#include <ostream>
+
+#pragma push_macro("_SCL_SECURE_NO_WARNINGS")
+#define _SCL_SECURE_NO_WARNINGS
+
+namespace azure { namespace storage_lite {
+
+    class memory_streambuf : public std::streambuf
+    {
+    public:
+        memory_streambuf(char* data, uint64_t size) : m_data(data), m_cur(data), m_end(data + size) {}
+
+    protected:
+        std::streamsize showmanyc() override
+        {
+            return m_end - m_cur;
+        }
+
+        int_type underflow() override
+        {
+            return traits_type::to_int_type(*m_cur);
+        }
+
+        int_type uflow() override
+        {
+            return traits_type::to_int_type(*m_cur++);
+        }
+
+        int_type overflow(int_type ch) override
+        {
+            *m_cur++ = traits_type::to_char_type(ch);
+            return ch;
+        }
+
+        std::streamsize xsgetn(char_type* s, std::streamsize count) override
+        {
+            std::streamsize read_count = std::min(count, showmanyc());
+            std::copy(m_cur, m_cur + read_count, s);
+            m_cur += read_count;
+            return read_count;
+        }
+
+        std::streamsize xsputn(const char_type* s, std::streamsize count) override
+        {
+            std::streamsize write_count = std::min(count, showmanyc());
+            std::copy(s, s + write_count, m_cur);
+            m_cur += write_count;
+            return write_count;
+        }
+
+        pos_type seekoff(off_type off, std::ios_base::seekdir dir, std::ios_base::openmode which) override
+        {
+            if (which & (std::ios_base::in | std::ios_base::out)) {
+                if (dir == std::ios_base::cur)
+                {
+                    m_cur += off;
+                }
+                else if (dir == std::ios_base::end)
+                {
+                    m_cur = m_end + off;
+                }
+                else if (dir == std::ios_base::beg)
+                {
+                    m_cur = m_data + off;
+                }
+            }
+            return m_cur - m_data;
```

Comment:
Is it by design to enable seek to an out-of-range position?

Reply:
No, out-of-range seeking is undefined bahavior.